### ISSUE TITLE
Implementation of pre trained nnUNet sct_deepseg models on Slicer

### ISSUE THREAD

**Issue by Nilser3** [spinalcordtoolbox/spinalcordtoolbox#4531]
# Description 

With [SlicerNNUnet](https://github.com/KitwareMedical/SlicerNNUnet) module is possible to implement [nnUNetV2](https://github.com/MIC-DKFZ/nnUNet) pre-trained models on [Slicer (v5.7.0)](https://download.slicer.org/). 
Currently, we have 6 models in `sct_deepseg` (see Table 01) trained with the nnUNetV2 framework that we could implement in Slicer by following these steps:

# Implementation 

1. Install [3D Slicer version 5.7.0](https://download.slicer.org/), then install the module[SlicerNNUnet](https://github.com/KitwareMedical/SlicerNNUnet) from the extensions explorer.

**Fig.01**
<img src="https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/77469192/9d7964d2-66e3-464d-ac1a-04caaaced63b" width="300px;" alt=""/>

2. Choose the model to implement, download from the SCT release page ( `.zip` file ~250 Mb) (see Table 01) and unzip.

**Table 01**
| Model | Release page |
|--|--|
|  seg_sc_lesion_t2w_sci | [Link  ](https://github.com/ivadomed/model_seg_sci/releases/)|
|  seg_spinal_rootlets_t2w | [Link  ](https://github.com/ivadomed/model-spinal-rootlets/releases/)|
|  seg_mouse_gm_wm_t1w | [Link  ](https://github.com/ivadomed/model_seg_mouse-sc_wm-gm_t1/releases/)|
|  seg_sc_ms_lesion_stir_psir |  [Link ](https://github.com/ivadomed/canproco/releases/)|
|  seg_sc_epi | [Link](https://github.com/sct-pipeline/fmri-segmentation/releases/)  |
|  seg_ms_lesion_mp2rage (on implementation) | [Link ](https://github.com/ivadomed/model_seg_ms_mp2rage/releases/)|

3.  Reorganization of downloaded files

Native arborescence on `sct_deepserg` model
```
spinalcordtoolbox/data/deepseg_models/model_seg_ms_lesion_mp2rage/
├── dataset_fingerprint.json
├── dataset.json
├── fold_3
│   ├── checkpoint_best.pth
│   ├── debug.json
│   ├── progress.png
│   └── training_log_2024_3_14_20_35_08.txt
├── plans.json
└── source.json 
``` 

Slicer friendly  arborescence 
```
Dataset403_seg_ms_lesion_mp2rage_1mm_322subj/
└── nnUNetTrainer_seg_ms_lesion_mp2rage_nnUNetPlans_3d_fullres
    ├── dataset_fingerprint.json
    ├── dataset.json
    ├── fold_3
    │   ├── checkpoint_final.pth
    │   ├── debug.json
    │   ├── progress.png
    │   └── training_log_2024_3_14_20_35_08.txt
    └── plans.json 
``` 
- Note 1: The arborescence is the same as the one generated in `nnUNet_results/` after training a model.
- Note 2: Verify that the name of Slicer friendly folder (`Dataset403_seg_ms_lesion_mp2rage_1mm_322subj` for us) is the same as in `plans.json` file (`dataset_name` flag).

4. nnUNet Install (See first row on Fig. 02): Follow the instructions

**Fig.02**
<img src="https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/77469192/498d57c6-e26b-415f-8ca1-aa61580474e6" width="500px;" alt=""/>

5. Implementation
- Navigate and select the Slicer friendly folder (See Fig.02) on `Model path`
- Folds: Choose the fold that is currently implemented (3 for us), however we could also implement the 5 trained folds.
- Preprocessing and pos processing steps (work in progress)
- Apply !! 🚀🚀

# Related issues
#4528

**Comment by joshuacwnewton** [comment]
Thank you so much for creating a very detailed and thorough explanation! 

> Note 2: Verify that the name of Slicer friendly folder (Dataset403_seg_ms_lesion_mp2rage_1mm_322subj for us) is the same into plans.json file (dataset_name flag).

I wonder if there is a way that we can update the releases of the models to make this easier. (e.g. changing `plans.json`, renaming/adding subfolders, etc.) so that it "just works" immediately after extracting the model folder.

**Comment by Nilser3** [comment]
Thanks you @joshuacwnewton ,
> 
> Right now, the user would be expected to do a few extra steps, which may confuse them/cause errors/etc.

ok, maybe if we create a `models box`  where the users can find our `sct_deepseg` nnUnet models (in Slicer friendly aborescence).
 
This could be also useful to quickly implement/share models that are not yet available in SCT version but that the community can test in Slicer.

**Comment by joshuacwnewton** [comment]
> (in Slicer friendly aborescence).

I'm curious -- Is the Slicer-friendly folder structure compatible with `sct_deepseg`? (In other words, if `sct_deepseg -install` tries to download the `.zip` that is structured for Slicer, will it still work? :thinking: 

Ideally, it would be nice if we could have a single `.zip` link that works for both Slicer and SCT. :thinking:

**Comment by joshuacwnewton** [comment]
Wonderful! We should be able to update the releases for each of the models to use this structure, I think?

(Tagging @valosekj as you've been involved in creating the `.zip` assets for most of SCT's new nnUNet models. Do you have any thoughts about `.zip` structure? :relaxed:)

**Comment by valosekj** [comment]
> Wonderful! We should be able to update the releases for each of the models to use this structure, I think?

Yes, definitely! Basically, it would just require adding two parent folders to our current structure:

<details><summary>current structure</summary>

```console
├── dataset.json
├── dataset_fingerprint.json
├── fold_all
│   ├── checkpoint_final.pth
│   ├── debug.json
│   ├── progress.png
│   ├── training_log_2024_1_9_15_28_27.png
│   └── training_log_2024_1_9_15_28_27.txt
├── plans.json
└── source.json
```

</details>

<details><summary>updated structure</summary>

```console
Dataset012_M5_2023-12-06
└── nnUNetTrainer_2000epochs__nnUNetPlans__3d_fullres
    ├── dataset.json
    ├── dataset_fingerprint.json
    ├── fold_all
    │   ├── checkpoint_final.pth
    │   ├── debug.json
    │   ├── progress.png
    │   ├── training_log_2024_1_9_15_28_27.png
    │   ├── training_log_2024_1_9_15_28_27.txt
    ├── plans.json
    └── source.json
```

</details>

And then some tiny update of the `sct_deepseg` inference script to handle these two extra folders will be required. Only note that these folder names will differ across models; but it should not be difficult to handle.

Tbh, I am trying to remember why we decided not to include the two extra folders. Maybe to avoid nested folders... @naga-karthik, don't you remember?

**Comment by naga-karthik** [comment]
I don't think there was any particular reason other than simply avoiding to deal with several folders to get to the final model checkpoint. If Slicer integration requires it then we can create a nested structure as well !

**Comment by valosekj** [comment]
Okay! Thank you for the confirmation! I believe we can then update our models! I can take care of [model_seg_sci](https://github.com/ivadomed/model_seg_sci/) and [model-spinal-rootlets](https://github.com/ivadomed/model-spinal-rootlets).
Is it better to create new releases or update assets (zip files) of the existing releases? I am in favor of creating new releases.
